File: native/photokit-helper/Sources/PhotoKitProtocol/Protocol.swift
import Foundation
import Photos

/// Protocol version shared by the native helper and Node client.
public let helperProtocolVersion = 1

/// Operations supported by the current helper protocol.
public enum ProtocolOperation: String, Codable, Sendable
{
    case authorizationRequest = "authorization-request"
    case authorizationStatus = "authorization-status"
    case version
}

/// Stable boundary-level errors returned by the helper.
public enum ProtocolErrorCode: String, Codable, Sendable
{
    case incompatibleProtocolVersion = "incompatible-protocol-version"
    case invalidRequest = "invalid-request"
    case nativeFailure = "native-failure"
    case unknownOperation = "unknown-operation"
}

/// A value representable in a JSON document.
public indirect enum JSONValue: Codable, Equatable, Sendable
{
    case array([JSONValue])
    case boolean(Bool)
    case null
    case number(Double)
    case object([String: JSONValue])
    case string(String)

    public init(from decoder: Decoder) throws
    {
        let container = try decoder.singleValueContainer()

        if container.decodeNil()
        {
            self = .null
        }
        else if let value = try? container.decode(Bool.self)
        {
            self = .boolean(value)
        }
        else if let value = try? container.decode(Double.self)
        {
            self = .number(value)
        }
        else if let value = try? container.decode(String.self)
        {
            self = .string(value)
        }
        else if let value = try? container.decode([JSONValue].self)
        {
            self = .array(value)
        }
        else
        {
            self = .object(try container.decode([String: JSONValue].self))
        }
    }

    public func encode(to encoder: Encoder) throws
    {
        var container = encoder.singleValueContainer()

        switch self
        {
        case .array(let value):
            try container.encode(value)
        case .boolean(let value):
            try container.encode(value)
        case .null:
            try container.encodeNil()
        case .number(let value):
            try container.encode(value)
        case .object(let value):
            try container.encode(value)
        case .string(let value):
            try container.encode(value)
        }
    }
}

/// Request decoded from the helper's single JSON argument.
public struct ProtocolRequestEnvelope: Codable, Equatable, Sendable
{
    public let operation: String
    public let parameters: [String: JSONValue]
    public let protocolVersion: Int

    public init(
        protocolVersion: Int = helperProtocolVersion,
        operation: String,
        parameters: [String: JSONValue] = [:]
    )
    {
        self.operation = operation
        self.parameters = parameters
        self.protocolVersion = protocolVersion
    }
}

/// Indicates that a peer uses a different helper protocol version.
public struct IncompatibleProtocolVersionError: Error, Equatable, Sendable
{
    public let expected: Int
    public let received: Int

    public init(expected: Int = helperProtocolVersion, received: Int)
    {
        self.expected = expected
        self.received = received
    }
}

/// Rejects a protocol version that the current helper cannot serve.
public func assertCompatibleProtocolVersion(_ received: Int) throws
{
    guard received == helperProtocolVersion else
    {
        throw IncompatibleProtocolVersionError(received: received)
    }
}

/// Successful response encoded by the helper.
public struct ProtocolSuccessEnvelope<Data: Codable & Equatable & Sendable>: Codable, Equatable, Sendable
{
    public let data: Data
    public let operation: String
    public let protocolVersion: Int
    public let success: Bool

    public init(operation: ProtocolOperation, data: Data)
    {
        self.data = data
        self.operation = operation.rawValue
        self.protocolVersion = helperProtocolVersion
        self.success = true
    }
}

/// Structured failure information encoded by the helper.
public struct ProtocolError: Codable, Equatable, Sendable
{
    public let code: ProtocolErrorCode
    public let details: [String: JSONValue]?
    public let message: String

    public init(code: ProtocolErrorCode, message: String, details: [String: JSONValue]? = nil)
    {
        self.code = code
        self.details = details
        self.message = message
    }
}

/// Failed response encoded by the helper.
public struct ProtocolFailureEnvelope: Codable, Equatable, Sendable
{
    public let error: ProtocolError
    public let operation: String?
    public let protocolVersion: Int
    public let success: Bool

    public init(operation: String?, error: ProtocolError)
    {
        self.error = error
        self.operation = operation
        self.protocolVersion = helperProtocolVersion
        self.success = false
    }

    private enum CodingKeys: String, CodingKey
    {
        case error
        case operation
        case protocolVersion
        case success
    }

    public init(from decoder: Decoder) throws
    {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.error = try container.decode(ProtocolError.self, forKey: .error)
        self.operation = try container.decodeIfPresent(String.self, forKey: .operation)
        self.protocolVersion = try container.decode(Int.self, forKey: .protocolVersion)
        self.success = try container.decode(Bool.self, forKey: .success)
    }

    public func encode(to encoder: Encoder) throws
    {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(error, forKey: .error)
        try container.encode(operation, forKey: .operation)
        try container.encode(protocolVersion, forKey: .protocolVersion)
        try container.encode(success, forKey: .success)
    }
}

/// Data returned by the version operation.
public struct HelperVersionData: Codable, Equatable, Sendable
{
    public let protocolVersion: Int

    public init(protocolVersion: Int = helperProtocolVersion)
    {
        self.protocolVersion = protocolVersion
    }
}

/// Photo library authorization states returned by PhotoKit.
public enum PhotoLibraryAuthorizationStatus: String, Codable, Sendable
{
    case authorized
    case denied
    case limited
    case notDetermined = "not-determined"
    case restricted
}

/// Data returned by authorization operations.
public struct AuthorizationStatusData: Codable, Equatable, Sendable
{
    public let canRequest: Bool
    public let guidance: String
    public let status: PhotoLibraryAuthorizationStatus

    public init(status: PhotoLibraryAuthorizationStatus)
    {
        self.canRequest = status == .notDetermined
        self.guidance = authorizationGuidance(for: status)
        self.status = status
    }
}

func authorizationGuidance(for status: PhotoLibraryAuthorizationStatus) -> String
{
    switch status
    {
    case .authorized:
        return "Photo library access is available."
    case .denied:
        return "Enable Photos access in System Settings > Privacy & Security > Photos."
    case .limited:
        return "Photo library access is limited to the selected assets."
    case .notDetermined:
        return "Send an authorization-request operation to ask for Photos access."
    case .restricted:
        return "Photos access is restricted by a system policy and cannot be requested."
    }
}

func photoLibraryAuthorizationStatus(from status: PHAuthorizationStatus) -> PhotoLibraryAuthorizationStatus
{
    switch status
    {
    case .authorized:
        return .authorized
    case .denied:
        return .denied
    case .limited:
        return .limited
    case .notDetermined:
        return .notDetermined
    case .restricted:
        return .restricted
    @unknown default:
        return .restricted
    }
}

/// Returns the current read/write authorization without prompting.
public func currentAuthorizationStatus() -> PhotoLibraryAuthorizationStatus
{
    photoLibraryAuthorizationStatus(from: PHPhotoLibrary.authorizationStatus(for: .readWrite))
}

func authorizationStatusAfterRequest(
    currentStatus: PhotoLibraryAuthorizationStatus,
    request: () async -> PhotoLibraryAuthorizationStatus
) async -> PhotoLibraryAuthorizationStatus
{
    guard currentStatus == .notDetermined else
    {
        return currentStatus
    }

    return await request()
}

/// Requests read/write authorization only while the status is undetermined.
public func requestPhotoLibraryAuthorization() async -> PhotoLibraryAuthorizationStatus
{
    await authorizationStatusAfterRequest(currentStatus: currentAuthorizationStatus())
    {
        await withCheckedContinuation
        { continuation in
            PHPhotoLibrary.requestAuthorization(for: .readWrite)
            { status in
                continuation.resume(returning: photoLibraryAuthorizationStatus(from: status))
            }
        }
    }
}

File: packages/core/src/protocol.ts
/** Protocol version shared by the Node wrapper and native helper. */
export const helperProtocolVersion = 1;

/** Operations supported by protocol version one. */
export const protocolOperations = ["version", "authorization-status", "authorization-request", "list-assets"] as const;

/** Stable errors returned by protocol version one. */
export const protocolErrorCodes = [
    "incompatible-protocol-version",
    "invalid-request",
    "native-failure",
    "photo-library-access-unavailable",
    "unknown-operation",
] as const;

/** A value representable in a JSON document. */
export type JsonValue = boolean | null | number | string | JsonValue[] | { [key: string]: JsonValue };

/** An operation supported by the current helper protocol. */
export type ProtocolOperation = typeof protocolOperations[number];

/** A stable helper protocol error code. */
export type ProtocolErrorCode = typeof protocolErrorCodes[number];

/** Request passed to the native helper as one JSON argument. */
export interface ProtocolRequestEnvelope
{
    operation: string;
    parameters: Record<string, JsonValue>;
    protocolVersion: number;
}

/** Structured failure information returned by the native helper. */
export interface ProtocolError
{
    code: string;
    details?: Record<string, JsonValue>;
    message: string;
}

/** Successful response returned by the native helper. */
export interface ProtocolSuccessEnvelope<TData = JsonValue>
{
    data: TData;
    operation: string;
    protocolVersion: number;
    success: true;
}

/** Failed response returned by the native helper. */
export interface ProtocolFailureEnvelope
{
    error: ProtocolError;
    operation: string | null;
    protocolVersion: number;
    success: false;
}

/** Response returned by the native helper. */
export type ProtocolResponseEnvelope<TData = JsonValue> = ProtocolFailureEnvelope | ProtocolSuccessEnvelope<TData>;

/** Photo library authorization states returned by the native helper. */
export type PhotoLibraryAuthorizationStatus = "authorized" | "denied" | "limited" | "not-determined" | "restricted";

/** Data returned by the native helper's version operation. */
export interface HelperVersionData
{
    protocolVersion: number;
}

/** Data returned by the native helper's authorization operations. */
export interface AuthorizationStatusData
{
    canRequest: boolean;
    guidance: string;
    status: PhotoLibraryAuthorizationStatus;
}

/** Image and video media types exposed by the helper. */
export type AssetMediaType = "image" | "video";

/** Stable names for PhotoKit media-subtype flags. */
export type AssetMediaSubtype = "photo-depth-effect" | "photo-hdr" | "photo-live" | "photo-panorama" | "photo-screenshot" | "spatial-media" | "video-cinematic" | "video-high-frame-rate" | "video-streamed" | "video-timelapse";

/** Parameters accepted by the list-assets operation. */
export interface ListAssetsParameters
{
    limit?: number;
    mediaType?: AssetMediaType;
}

/** Metadata returned for one PhotoKit asset without requesting its content. */
export interface AssetMetadata
{
    creationDate: string | null;
    duration: number | null;
    favorite: boolean;
    hidden: boolean;
    localIdentifier: string;
    mediaSubtypes: AssetMediaSubtype[];
    mediaType: AssetMediaType;
    modificationDate: string | null;
    pixelHeight: number;
    pixelWidth: number;
}

/** Data returned by the list-assets operation. */
export interface AssetListData
{
    assets: AssetMetadata[];
}

/** Indicates that a helper response does not match the protocol envelope. */
export class InvalidProtocolResponseError extends Error
{
    public constructor(message: string, options?: ErrorOptions)
    {
        super(message, options);
        this.name = "InvalidProtocolResponseError";
    }
}

/** Indicates that the native helper uses a different protocol version. */
export class IncompatibleProtocolVersionError extends Error
{
    public readonly expected: number;
    public readonly received: number;

    public constructor(expected: number, received: number)
    {
        super(`Expected helper protocol version ${expected}, received ${received}.`);
        this.name = "IncompatibleProtocolVersionError";
        this.expected = expected;
        this.received = received;
    }
}

function isRecord(value: unknown): value is Record<string, unknown>
{
    return typeof value === "object" && value !== null && !Array.isArray(value);
}

function isJsonValue(value: unknown): value is JsonValue
{
    if (value === null || ["boolean", "number", "string"].includes(typeof value))
    {
        return typeof value !== "number" || Number.isFinite(value);
    }

    if (Array.isArray(value))
    {
        return value.every(isJsonValue);
    }

    return isRecord(value) && Object.values(value).every(isJsonValue);
}

function decodeError(value: unknown): ProtocolError
{
    if (!isRecord(value) || typeof value.code !== "string" || typeof value.message !== "string")
    {
        throw new InvalidProtocolResponseError("Helper failure is missing a structured error.");
    }

    if (value.details !== undefined && (!isRecord(value.details) || !isJsonValue(value.details)))
    {
        throw new InvalidProtocolResponseError("Helper failure contains invalid error details.");
    }

    return {
        code: value.code,
        details: value.details,
        message: value.message,
    };
}

/** Encodes one operation as the native helper's JSON request argument. */
export function encodeProtocolRequest(
    operation: ProtocolOperation,
    parameters: Record<string, JsonValue> = {},
): string
{
    const request: ProtocolRequestEnvelope = {
        operation,
        parameters,
        protocolVersion: helperProtocolVersion,
    };

    return JSON.stringify(request);
}

/** Decodes and validates one JSON response from the native helper. */
export function decodeProtocolResponse(input: string): ProtocolResponseEnvelope
{
    let value: unknown;

    try
    {
        value = JSON.parse(input);
    }
    catch (error)
    {
        throw new InvalidProtocolResponseError("Helper response is not valid JSON.", { cause: error });
    }

    if (!isRecord(value) || !Number.isInteger(value.protocolVersion))
    {
        throw new InvalidProtocolResponseError("Helper response is missing an integer protocolVersion.");
    }

    if (value.protocolVersion !== helperProtocolVersion)
    {
        throw new IncompatibleProtocolVersionError(helperProtocolVersion, value.protocolVersion as number);
    }

    if (value.success === true && typeof value.operation === "string" && isJsonValue(value.data))
    {
        return {
            data: value.data,
            operation: value.operation,
            protocolVersion: value.protocolVersion as number,
            success: true,
        };
    }

    if (value.success === false && (typeof value.operation === "string" || value.operation === null))
    {
        return {
            error: decodeError(value.error),
            operation: value.operation,
            protocolVersion: value.protocolVersion as number,
            success: false,
        };
    }

    throw new InvalidProtocolResponseError("Helper response does not match a success or failure envelope.");
}
